### PR TITLE
Logs optimization

### PR DESCRIPTION
**Pull request by EstherDarkish**
New API endpoints for report-server.

### CODE REVIEW

File: src/test/resources/expectedReporterAndElements.json
Comment (Tristan-WorkGH) [inline src/test/resources/expectedReporterAndElements.json:1]: Did we add the "ID" field in reporters to check they are added by `buildTreeFromReportersAndElements` ?

File: src/main/java/org/gridsuite/report/server/repositories/TreeReportRepository.java [hunk 1]
```diff
@@ -23,13 +25,34 @@ public interface TreeReportRepository extends JpaRepository<TreeReportEntity, UU
 
     List<TreeReportEntity> findAllByReportId(UUID uuid);
 
+    @EntityGraph(attributePaths = {"values", "dictionary"}, type = EntityGraph.EntityGraphType.LOAD)
     List<TreeReportEntity> findAllByParentReportIdNode(UUID uuid);
```
Comment (Tristan-WorkGH) [inline src/main/java/org/gridsuite/report/server/repositories/TreeReportRepository.java:29]: unused method
```suggestion
```

File: src/main/java/org/gridsuite/report/server/repositories/ReportElementRepository.java [hunk 2]
```diff
@@ -21,8 +22,12 @@
 @Repository
 public interface ReportElementRepository extends JpaRepository<ReportElementEntity, UUID> {
 
+    @EntityGraph(attributePaths = {"values"}, type = EntityGraph.EntityGraphType.LOAD)
     List<ReportElementEntity> findAllByParentReportIdNode(UUID uuid);
```
Comment (Tristan-WorkGH) [inline src/main/java/org/gridsuite/report/server/repositories/ReportElementRepository.java:26]: unused method
```suggestion
```

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 3]
```diff
@@ -147,15 +232,26 @@ private ReportValueEmbeddable toValueEmbeddable(Map.Entry<String, TypedValue> en
         return new ReportValueEmbeddable(entryValue.getKey(), entryValue.getValue().getValue(), entryValue.getValue().getType());
     }
 
+    private String maxSeverity(ReporterModel reporter) {
```
Comment (Tristan-WorkGH) [inline src/main/java/org/gridsuite/report/server/ReportService.java:235]:
```suggestion
    private static String maxSeverity(ReporterModel reporter) {
```
can be static method

File: src/test/resources/expectedReporterAndElements.json
Comment (EstherDarkish) [reply src/test/resources/expectedReporterAndElements.json:1]: If we add them, the generated UUIDs will always be different

File: src/main/java/org/gridsuite/report/server/repositories/TreeReportRepository.java (hunk 1 above)
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/repositories/TreeReportRepository.java]: Thanks, removed.

File: src/main/java/org/gridsuite/report/server/repositories/ReportElementRepository.java (hunk 2 above)
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/repositories/ReportElementRepository.java]: Thanks, removed.

File: src/main/java/org/gridsuite/report/server/ReportService.java (hunk 3 above)
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/ReportService.java]: Thanks, updated.

**Review by Tristan-WorkGH — APPROVED** [review]

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 4]
```diff
@@ -56,51 +72,118 @@ public ReportService(final ReportRepository reportRepository, TreeReportReposito
...
+    public ReporterModel getReporterStructureAndElements(UUID reporterId, Set<String> severityLevels) {
+        Objects.requireNonNull(reporterId);
+        TreeReportEntity treeReportEntity = treeReportRepository.findById(reporterId).orElseThrow(EntityNotFoundException::new);
+
+        var report = new ReporterModel(treeReportEntity.getIdNode().toString(), treeReportEntity.getIdNode().toString());
+        report.addSubReporter(getTreeReportAndDescendantElements(treeReportEntity, true, severityLevels));
+        return report;
+    }
+
+    ReporterModel getTreeReportAndDescendantElements(TreeReportEntity treeReportEntity, boolean getElements, Set<String> severityLevels) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/report/server/ReportService.java:126]: Make method `private` and rename by `getTreeReport`
Rename `getElements` by withElements

File: src/main/java/org/gridsuite/report/server/ReportController.java [hunk 5]
```diff
@@ -36,29 +37,60 @@ public ReportController(ReportService service) {
...
-            return errorOnReportNotFound ? ResponseEntity.notFound().build() : ResponseEntity.ok().body(service.getEmptyReport(id, defaultName).getSubReporters());
+            return ResponseEntity.ok().body(service.getEmptyReport(id, defaultName).getSubReporters());
+        }
+    }
+
+    @GetMapping(value = "/reports/{id}/elements", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Get the elements of a report, its reporters, and their subreporters")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The elements of the report, reporters and subreporters"),
+        @ApiResponse(responseCode = "404", description = "The report does not exists")})
+    public ResponseEntity<List<ReporterModel>> getReportStructureAndElements(@PathVariable("id") UUID id,
```
[the remark this answers is not in the document]
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/ReportController.java]: Done

File: src/main/java/org/gridsuite/report/server/ReportController.java [hunk 6]
```diff
@@ -36,29 +37,60 @@ public ReportController(ReportService service) {
...
-    @GetMapping(value = "reports/{id}", produces = MediaType.APPLICATION_JSON_VALUE)
-    @Operation(summary = "Get report by id")
-    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The report"),
-        @ApiResponse(responseCode = "404", description = "The report does not exists")})
-    public ResponseEntity<List<ReporterModel>> getReport(@PathVariable("id") UUID id,
-                                                   @Parameter(description = "Return 404 if report is not found or empty report") @RequestParam(name = "errorOnReportNotFound", required = false, defaultValue = "true") boolean errorOnReportNotFound,
+    @GetMapping(value = "reports/{id}/reporters", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Get the tree structure of a report and its reporters")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The tree structure of a report and its reporters")})
+    public ResponseEntity<List<ReporterModel>> getReportStructure(@PathVariable("id") UUID id,
```
[the remark this answers is not in the document]
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/ReportController.java]: Done.

File: src/main/java/org/gridsuite/report/server/repositories/TreeReportRepository.java [hunk 7]
```diff
@@ -23,13 +25,31 @@ public interface TreeReportRepository extends JpaRepository<TreeReportEntity, UU
 
     List<TreeReportEntity> findAllByReportId(UUID uuid);
 
-    List<TreeReportEntity> findAllByParentReportIdNode(UUID uuid);
+    @EntityGraph(attributePaths = {"values", "dictionary"}, type = EntityGraph.EntityGraphType.LOAD)
```
[the remark this answers is not in the document]
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/repositories/TreeReportRepository.java:28]: Done

File: src/main/java/org/gridsuite/report/server/repositories/ReportElementRepository.java [hunk 8]
```diff
@@ -21,7 +22,8 @@
 @Repository
 public interface ReportElementRepository extends JpaRepository<ReportElementEntity, UUID> {
 
-    List<ReportElementEntity> findAllByParentReportIdNode(UUID uuid);
+    @EntityGraph(attributePaths = {"values"}, type = EntityGraph.EntityGraphType.LOAD)
```
[the remark this answers is not in the document]
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/repositories/ReportElementRepository.java:25]: Done.

File: src/main/java/org/gridsuite/report/server/ReportController.java [hunk 9]
```diff
@@ -36,29 +37,60 @@ public ReportController(ReportService service) {
...
+        } catch (EntityNotFoundException ignored) {
+            return ResponseEntity.notFound().build();
+        }
+    }
+
+    @GetMapping(value = "/reporters/{id}/elements", produces = MediaType.APPLICATION_JSON_VALUE)
+    @Operation(summary = "Get the elements of a reporter and its subreporters")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The elements of the reporter and its subreporters"),
+        @ApiResponse(responseCode = "404", description = "The reporter does not exists")})
+    public ResponseEntity<List<ReporterModel>> getReporterStructureAndElements(@PathVariable("id") UUID id,
```
[the remark this answers is not in the document]
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/ReportController.java]: Done.

File: src/main/java/org/gridsuite/report/server/ReportService.java (hunk 4 above)
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/ReportService.java:113]: Done

File: src/main/java/org/gridsuite/report/server/ReportService.java [hunk 10]
```diff
@@ -56,51 +72,112 @@ public ReportService(final ReportRepository reportRepository, TreeReportReposito
...
+        }
+
+        // We need to get the entities to have access to the dictionaries
+        List<TreeReportEntity> treeReportEntities = treeReportRepository.findAllByIdNodeInOrderByNanos(treeReportEntitiesIds);
+
+        // Now we can rebuild the tree
+        return buildTreeFromReportersAndElements(treeReportEntity, treeReportEntities, allReportElements);
+    }
+
+    private ReporterModel buildTreeFromReportersAndElements(final TreeReportEntity rootTreeReportEntity, final List<TreeReportEntity> allTreeReports,
```
[the remark this answers is not in the document]
Comment (EstherDarkish) [reply src/main/java/org/gridsuite/report/server/ReportService.java:136]: Done

File: src/test/java/org/gridsuite/report/server/TreeReportRepositoryTest.java [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @After
+    public void tearOff() {
+        // clean DB
+        treeReportRepository.deleteAll();
+    }
+
+    @Test
+    public void testCreateTreeReport() {
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/report/server/TreeReportRepositoryTest.java:62]: Why we do not an integration test (create a report) in ControllerTest

**Review by SlimaneAmar — APPROVED** [review]

File: src/test/java/org/gridsuite/report/server/TreeReportRepositoryTest.java (hunk 11 above)
Comment (EstherDarkish) [reply src/test/java/org/gridsuite/report/server/TreeReportRepositoryTest.java:62]: Yeah, maybe this file is not useful ?

**Review by dbraquart — APPROVED** [review]
code: ok
tests: ok

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**